server/types/array.go

```diff
@@ -0,0 +1,433 @@
+	return types.ExtendedTypeSerializedWidth_Unbounded
+}
+
+// MaxTextResponseByteLength implements the DoltgresType interface.
+func (ac arrayContainer) MaxTextResponseByteLength(ctx *sql.Context) uint32 {
+	return math.MaxUint32
+}
+
+// OID implements the DoltgresType interface.
+func (ac arrayContainer) OID() uint32 {
+	return uint32(ac.oid)
+}
+
+// Promote implements the DoltgresType interface.
+func (ac arrayContainer) Promote() sql.Type {
+	return ac
+}
+
+// SerializedCompare implements the DoltgresType interface.
+func (ac arrayContainer) SerializedCompare(v1 []byte, v2 []byte) (int, error) {
+	//TODO: write a far more optimized version of this that does not deserialize the entire arrays at once
+	dv1, err := ac.DeserializeValue(v1)
+	if err != nil {
+		return 0, err
+	}
+	dv2, err := ac.DeserializeValue(v2)
+	if err != nil {
+		return 0, err
+	}
+	return ac.Compare(dv1, dv2)
+}
+
+// SerializeType implements the DoltgresType interface.
+func (ac arrayContainer) SerializeType() ([]byte, error) {
+	innerSerialized, err := ac.innerType.SerializeType()
+	if err != nil {
+		return nil, err
+	}
+	serialized := make([]byte, len(innerSerialized)+2)
+	binary.LittleEndian.PutUint16(serialized, uint16(ac.serializationID))
+	copy(serialized[2:], innerSerialized)
+	return serialized, nil
+}
+
+// SQL implements the DoltgresType interface.
+func (ac arrayContainer) SQL(ctx *sql.Context, dest []byte, valInterface any) (sqltypes.Value, error) {
+	return ac.funcs.SQL(ctx, ac, dest, valInterface)
+}
+
+// String implements the DoltgresType interface.
+func (ac arrayContainer) String() string {
+	return ac.innerType.String() + "[]"
+}
+
+// ToArrayType implements the DoltgresType interface.
+func (ac arrayContainer) ToArrayType() DoltgresArrayType {
+	return ac
+}
+
+// Type implements the DoltgresType interface.
+func (ac arrayContainer) Type() query.Type {
+	return sqltypes.Text
+}
+
+// ValueType implements the DoltgresType interface.
+func (ac arrayContainer) ValueType() reflect.Type {
+	return reflect.TypeOf([]any{})
+}
+
+// Zero implements the DoltgresType interface.
+func (ac arrayContainer) Zero() any {
+	return []any{}
+}
+
+// SerializeValue implements the DoltgresType interface.
+func (ac arrayContainer) SerializeValue(valInterface any) ([]byte, error) {
+	// Check for a nil value and convert to the expected type
+	if valInterface == nil {
+		return nil, nil
+	}
+	converted, _, err := ac.Convert(valInterface)
+	if err != nil {
+		return nil, err
+	}
+	vals := converted.([]any)
+
+	// Create the buffer that we'll write the array's contents to
+	arrayBuffer := bytes.Buffer{}
+	innerSerializedWidth := ac.innerType.MaxSerializedWidth()
+	// Write the total length to a buffer. We'll reuse this buffer for all uint-to-bytes operations.
+	lengthBuffer := make([]byte, 8)
+	binary.BigEndian.PutUint64(lengthBuffer, uint64(len(vals)))
+	arrayBuffer.Write(lengthBuffer)
+
+	// Each value is serialized as the following: IsNull, Size, Data
```

Comment:
Fine to check in, but before finalizing this you might want to consider something closer to Dolt's row storage format, with absolute offsets stored at the end, rather than [size,data] tuples. The problem with this approach is it's O(N) to access a random element in the middle, because you have to read up to that point.